pkg/controller/endpoint/endpoints_controller_cache.go

```
@@ -0,0 +1,234 @@
+implementing the network programming latency metric.
+
+The cache should be initialized via pod and service listers and kept up-to-date via the
+(Add|Update|Delete)(Pod|Service) methods called in the EndpointsController's event handlers.
+The Initialize method together with the event handlers method are guarded by the same mutex which
+guarantees that no information is ever missed and all events are processed in the serial order.
+
+The cache doesn't store full objects, it stores only references to minimize the memory
+consumption.
+ */
+type endpointsControllerCache struct {
+	pods map[Namespace]map[Name]*v1.Pod
+	services map[Namespace]map[Name]*v1.Service
+
+	// Mutex guarding this util.
+	mutex sync.Mutex
+}
+
+// String type wrappers to make it clear how the objects are indexed in maps.
+type Namespace string
+type Name string
+
+// Creates new instance of the endpointsControllerCache.
+func newEndpointsControllerCache() *endpointsControllerCache {
+	return &endpointsControllerCache{
+		pods : make(map[Namespace]map[Name]*v1.Pod),
+		services : make(map[Namespace]map[Name]*v1.Service),
+	}
+}
+
+// Initializes the cache with pods and services obtained from the provided listers.
+// Pods and services are listed behind mutex, which temporarily disables event listener callbacks
+// (e.g. AddPod, UpdateService, etc.). It guarantees that no information will be lost.
+// This method should be called exactly once on the startup, before the first call of the
+// GetServiceAndPods method, after the lister have synced.
+func (e *endpointsControllerCache) Initialize(serviceLister corelisters.ServiceLister, podLister corelisters.PodLister) {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	services, err := serviceLister.List(labels.Everything())
+	if err != nil {
+		// Error shouldn't happened as we're getting objects from a local cache.
+		utilruntime.HandleError(fmt.Errorf("Unable to list services due to %v", err))
```

errors in golang should start with small letter (here and below).

that said, while the error shouldn't happen, we shouldn't silently ignore it - it's safer to fail the initialization than leave the controller in partially broken state.

---

Done, let me know if klog.Fatal is not the right way to do that.